feat:  Design and implement several variable charts

wj23027 opened the pull request:
## Brief Information

This pull request is in the type of ([more info](https://github.com/angular/angular/blob/main/CONTRIBUTING.md#type) about types):

- [ ] build
- [ ] ci
- [ ] docs
- [x] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] test

Related issues ([all available keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

- resolve #654 

## Details
<!-- What did you do in this PR?  -->
I have implemented a basic grid layout for the charts, allowing different-sized charts to be added to the layout. 

![image](https://github.com/hypertrons/hypertrons-crx/assets/62098023/ef30e719-e1e8-438c-b55e-48c66ff6fd30)

In the next steps, I will need to add logic to allow users to add charts of specific sizes and types to specific positions, and optimize the styling of the layout.

Next Steps:
- [ ] Adjust the visual presentation of the charts.
- [ ] Implement Additional Chart Styles
- [ ] optimize the styling of the layout
- [ ] add logic to allow users to add charts of specific sizes and types to specific positions
- [ ] i18n
- [ ] Integrate the feature into 'Repo Collections' feature
## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/hypertrons/hypertrons-crx/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/hypertrons/hypertrons-crx)

## Others
<!-- Other information you want to share.  -->
<!-- If none to share, leave an "N.A."  -->

andyhuang18 commented:
Great! Wish you have a good time in your trip. 😄

wxharry commented:
Hi, @wj23027 . I have changed this base branch to `feat/repo-collection` since you and Andy is working on one feature. I have implmented a modal view for testing new charts. It helps to focus on developing the component itself. Since this branch highly depends on the work from Andy, my suggestion is that you can work on new charts first, wait til Andy implements the modal.

tyn1998 reviewed (changes requested):
Hi @wj23027, I found that the GridLayout is not imported by any file so I guess the screenshot you provided in the PR description is just for preview. Since @wxharry [provided a way ](https://github.com/hypertrons/hypertrons-crx/pull/716#issuecomment-1690978483)to test your charts, you can import GridLayout there.

Besides, please comment the code in English and hurry up :D

wj23027 commented:
This update introduces two new charts: boxplot chart and a Sankey chart . The box plot is used to illustrate the approximate distribution of issue response times, while the Sankey chart showcases user activity across different repositories.

boxplot chart :

![image](https://github.com/hypertrons/hypertrons-crx/assets/62098023/fae39feb-455a-4fd2-be07-2494a49640d0)

Sankey chart：

![image](https://github.com/hypertrons/hypertrons-crx/assets/62098023/d89067b2-2d18-4d95-b9eb-4c356819335f)

wj23027 commented:
In addition, a simple grid layout component has been implemented to display the charts separately in both the modal and option pages.

![image](https://github.com/hypertrons/hypertrons-crx/assets/62098023/463cb4cf-090d-4376-9627-d0c89fc60d67)

Furthermore, clicking on menu items can highlight data for corresponding repositories in certain charts.

![1695573083866](https://github.com/hypertrons/hypertrons-crx/assets/62098023/f0e7e103-5677-4a03-9ad0-22a0936e3f31)

Next Tasks :
- Implement highlighting for other charts.
- Fine-tune chart styles.
- Merge Andy's code contributions.

tyn1998 commented:
@wj23027 The Sankey chart is really good.

There are only two repos in the demo collection, could you add more repos into it? Please bare in mind that these charts should be scalable for any given collection of different size.

tyn1998 commented:
> @wj23027 The Sankey chart is really good.
> 
> There are only two repos in the demo collection, could you add more repos into it? Please bare in mind that these charts should be scalable for any given collection of different size.

Hello @wj23027, could you add more repos into your mock data? Since in my view, these two charts are not a good choice for those collections with 3+ repos.

<img width="1682" alt="image" src="https://github.com/hypertrons/hypertrons-crx/assets/32434520/a85ba0fb-4d15-4d60-b781-fa02a9bcc3df">

tyn1998 commented on src/pages/Options/variable-charts/GridLayout.tsx:
```diff
@@ -0,0 +1,56 @@
+            key={index}
+            className={`grid-item size-${chartSize[size]}`}
+            onClick={() => setGridItemSize(index, newsize)}
+          >
+            {/* 图表组件，根据占用情况来决定显示的内容 */}
+            {size > 0 ? <LineChart /> : ''}
+          </div>
+        );
```
Hello @wj23027, There is an error in line#40, if you `yarn run start` you'll find it.

<img width="854" alt="image" src="https://github.com/hypertrons/hypertrons-crx/assets/32434520/f8b7e6ed-f21a-40a9-bd1b-4938cb85f6c9">

wj23027 replied:
The error likely resulted from missing required parameters. It's a deprecated file, which I've now removed.

tyn1998 commented on src/helpers/get-newest-month.ts:
```diff
@@ -0,0 +1,11 @@
+const getNewestMonth = () => {
+  const lastDataAvailableMonth = new Date();
+  lastDataAvailableMonth.setDate(0);
+  return (
+    lastDataAvailableMonth.getFullYear() +
+    '-' +
+    (lastDataAvailableMonth.getMonth() + 1).toString().padStart(2, '0')
+  );
+};
+
+export default getNewestMonth;
```
This way to get the lastDataAvailableMonth is not robust. A solid way is implemented in #635, which you can refer to :)

wj23027 replied:
Thank you for pointing it out! I've made the necessary changes!

tyn1998 reviewed (changes requested):
Hi @wj23027, thanks for your time and effort contributed to this OSPP project. Here are some suggestions for you and I will give more of them in the following days.

wj23027 commented:
In this update, I added numerical display panels.I tried increasing the number of repositories in the collection and noticed that after the increase, the code additions and deletions chart as well as the multi-dimensional bar chart didn't look very aesthetically pleasing. So, I made the following adjustments: 
1.  Changed the code additions and deletions chart to an area chart,
2.  Limited the display range of the multi-dimensional bar chart to the last 6 months, 
3.  Added stacked bar chart.

wj23027 commented:
Here is a screenshot of the updated charts:

![image](https://github.com/hypertrons/hypertrons-crx/assets/62098023/de7a3fdb-93e5-41f4-b28a-8673650351fa)

![image](https://github.com/hypertrons/hypertrons-crx/assets/62098023/53a9c67a-f24f-48c7-83be-8184cd63e962)

wj23027 commented:
Additionally, I've noticed that when the collection contains too many developers, there are some issues with the Sankey chart's display. Therefore, I've increased the size of the Sankey chart, but it still works best for collections containing no more than 150 developers (ideally, fewer than 100).

wj23027 commented:
I have an issue to address: in the [stacked bar chart](https://github.com/hypertrons/hypertrons-crx/pull/716/files#diff-58ce18998c0f216466c14d33a617a8ff12ea43ea29d3d13e5978cc33283d8776), the data from different series isn't stacking as intended, but I haven't found the reason yet.

![image](https://github.com/hypertrons/hypertrons-crx/assets/62098023/1538b8c4-c8b4-4201-b67d-a925f79b0f15)

andyhuang18 commented:
> Here is a screenshot of the updated charts:
> 
> ![image](https://user-images.githubusercontent.com/62098023/275334398-de7a3fdb-93e5-41f4-b28a-8673650351fa.png)
> 
> ![image](https://user-images.githubusercontent.com/62098023/275334432-53a9c67a-f24f-48c7-83be-8184cd63e962.png)

Looks good!

wxharry commented:
> Additionally, I've noticed that when the collection contains too many developers, there are some issues with the Sankey chart's display. Therefore, I've increased the size of the Sankey chart, but it still works best for collections containing no more than 150 developers (ideally, fewer than 100).

Great! I am glad you tested it. Then we should add a warning or notice like when users are trying to add a sankey chart. I prefer adding it anyway since this is a very cool chart. What do you think @tyn1998 ?



> I have an issue to address: in the [stacked bar chart](https://github.com/hypertrons/hypertrons-crx/pull/716/files#diff-58ce18998c0f216466c14d33a617a8ff12ea43ea29d3d13e5978cc33283d8776), the data from different series isn't stacking as intended, but I haven't found the reason yet.
> 

I think it is because the two arries of data have different length.
<img width="374" alt="image" src="https://github.com/hypertrons/hypertrons-crx/assets/39271899/bf4fcbcd-e223-4cc5-aa87-f7d6d0420fc3">

wxharry commented:
Also, please pay attention to the style below. The charts are not supposed to be outside of their containers. I hope it's a small fix, but if it's not, a temporary workaround is acceptable.
<img width="960" alt="image" src="https://github.com/hypertrons/hypertrons-crx/assets/39271899/7ce8fadb-e0ea-47a3-b307-8b6f23361e06">

zhicheng-ning commented:
Some suggestions：
- The dimensions of the chart need to be adapted. For example：change it from150000 to 150k
- The maximum value of the chart is adapted,it will be more beautiful. For example：change maximun value from 120 to 20（The maximum value among all existing values）.

> Also, please pay attention to the style below. The charts are not supposed to be outside of their containers. I hope it's a small fix, but if it's not, a temporary workaround is acceptable. <img alt="image" width="960" src="https://user-images.githubusercontent.com/39271899/276479179-7ce8fadb-e0ea-47a3-b307-8b6f23361e06.png">

wj23027 commented:
Hi, I merged the collection contents(including the repolist and dashboard) into the Collection Modal~

![image](https://github.com/hypertrons/hypertrons-crx/assets/62098023/9db26e4f-05d4-4e2b-9d1e-188bc10f2276)

tyn1998 reviewed (approved):
Great job @wj23027  !

Your work LGTM, thank you for your contributions!
